blog/tech-talk-window-resize-behavior.md

```diff
@@ -0,0 +1,290 @@
+
+But because we had no draw instructions for the 400 pixels on the right, those areas didn't get updated.
+
+On the first resize, we'd show black pixels there. (That's the color that we initialize the surface with.)
+
+On subsequent resizes, those areas would show whatever was drawn to them before. We'd see stale pixels.
+
+I landed a fix for this issue in [crrev.com/c/7156576](https://crrev.com/c/7156576).
+
+The fix changes what we do when we receive a frame with a different size than our window. Instead of resizing the frame and adding gutter that contains stale pixels, we [resize our viewport](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/frame_sinks/root_compositor_frame_sink_impl.cc;l=555-556;drc=78bd63326c0a2489e6b016140aa4d84014af077a) and our clip rect.
+
+![Visualization where the clip rect and viewport size are adjusted to the frame size](/assets/img/blog/tech-talk-window-resize-behavior/visualization-6.png)
+
+We clip our surface to the size of the frame that we received. We don't show anything beyond the 600 pixels that we have draw instructions for.
+
+Voilà, no more gutter, no more stale pixels!
+
+(Side note: Without `supports_viewporter`, this would be an expensive operation because it would allocate a new output surface. However, with DirectComposition, we use the "viewporter" feature. So we don't reallocate the surface when we change the viewport size. We just make a different portion of it visible. Thus, it is a cheap operation.)
+
+## Backporting the patches to Electron
+
+Once the fixes made it into Chromium, we had to pull them into Electron, too.
+
+On the `main` branch, Electron updates its Chromium version constantly. As a result, the patches were merged into `main` in a [Chromium roll PR](https://github.com/electron/electron/pull/49145).
+
+However, commits that make it into `main` right now will only be included in an Electron release in about three months. Our existing release and pre-release branches run on older Chromium versions.
+
+The Chromium patches made it into Chromium 145. Electron 39 runs on Chromium 142. The upcoming Electron 40 will run on Chromium 144. (For comparison, the stable channel of Google Chrome is based on Chromium 143 right now.)
+
+Thus, the next step was to backport the patches to [Electron 39](https://github.com/electron/electron/pull/49138) and [Electron 40](https://github.com/electron/electron/pull/49191).
+
+Electron keeps a list of Chromium patches in the [`patches/chromium` directory](https://github.com/electron/electron/tree/a90ccc753b4d38266323dd054db2b98a45917117/patches/chromium). When we backport a Chromium patch, we add it there. When building Electron, these patches are applied to the Chromium source code.
+
+(In general, we try to keep the number of Chromium patches low. Every patch can lead to merge conflicts during Chromium updates. The maintenance burden from patches is real.)
```

Comment:
This would be a good place to link directly to the docs and also clarify that the patch is both justified and temporary.

```suggestion
In general, we [try to keep the number of Chromium patches low](https://www.electronjs.org/docs/latest/development/patches). Patches lead to merge conflicts during Chromium updates, and the maintenance burden is real.

In this case, the significant performance improvement justified that burden. It also helps to know that the patch is temporary and will soon be addressed upstream.
```